lightning/src/routing/router.rs

```diff
@@ -242,16 +248,96 @@ impl_writeable_tlv_based!(RouteHop, {
 	(10, cltv_expiry_delta, required),
 });
 
+/// The blinded portion of a [`Path`], if we're routing to a recipient who provided blinded paths in
+/// their BOLT12 [`Invoice`].
+///
+/// [`Invoice`]: crate::offers::invoice::Invoice
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct BlindedTail {
+	/// The blinded path itself, provided by the recipient.
+	pub path: BlindedPath,
+	/// The total fee paid on this blinded path, calculated using the [`BlindedPayInfo`] provided in
+	/// the BOLT 12 invoice.
+	///
+	/// [`BlindedPayInfo`]: crate::offers::invoice::BlindedPayInfo
+	pub fee_msat: u64,
+	/// Number of blocks subtracted from an incoming HTLC's `cltv_expiry` for the entire blinded path.
+	pub cltv_expiry_delta: u32,
+	/// The total amount paid on this [`Path`], excluding the fees.
+	pub final_value_msat: u64,
+	/// The scid used to reach the introduction node of the [`BlindedPath`].
+	///
+	/// [`BlindedPath`]: crate::blinded_path::BlindedPath
+	pub intro_node_scid: u64,
+}
+
+impl_writeable_tlv_based!(BlindedTail, {
+	(0, path, required),
+	(2, fee_msat, required),
+	(4, cltv_expiry_delta, required),
+	(6, final_value_msat, required),
+	(8, intro_node_scid, required),
+});
+
+/// A path in a [`Route`] to the payment recipient.
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+pub struct Path {
+	/// The list of unblinded hops in this [`Path`]. May be empty if a blinded path is present,
+	/// otherwise the last hop is the destination and this must be at least length one.
+	pub hops: Vec<RouteHop>,
+	/// The blinded path at which this path terminates, if we're sending to one, and its metadata.
+	pub blinded_tail: Option<BlindedTail>,
+}
+
+impl Path {
+	/// Returns the total amount of fees paid on this [`Path`].
+	pub fn fee_msat(&self) -> u64 {
+		match &self.blinded_tail {
+			Some(tail) => self.hops.iter().map(|hop| hop.fee_msat).sum::<u64>() + tail.fee_msat,
+			None =>
+				self.hops.split_last().map_or(0,
+					|(_, path_prefix)| path_prefix.iter().map(|hop| hop.fee_msat).sum())
+			}
+	}
+
+	/// Returns the total amount paid on this [`Path`], excluding the fees.
+	pub fn final_value_msat(&self) -> u64 {
+		match &self.blinded_tail {
+			Some(blinded_tail) => blinded_tail.final_value_msat,
+			None => self.hops.last().map_or(0, |hop| hop.fee_msat)
+		}
+	}
+
+	/// The length of this path, including the hops of its [`Path::blinded_tail`], if any.
+	pub fn len(&self) -> usize {
+		self.hops.len() + self.blinded_tail.as_ref().map_or(0, |tail| tail.path.blinded_hops.len())
```

Comment:
Will this double count the hop to the introduction node?

Reply:
It shouldn't?

Reply:
Does the last element of `self.hops` include the channel to the introduction node? Or is that accounted for by the first element of `self.blinded_tail`?

Reply:
The latter. The channel to the intro node is stored in `BlindedTail::intro_node_scid`

Reply:
Ah, so even though `self.hops` is the number of channels in the unblinded portion excluding `BlindedTail::intro_node_scid`, the length of the blinded portion is essentially the number of hops in it plus one. (i.e., the first element accounts for the channel with `BlindedTail::intro_node_scid`)

Reply:
I think that makes sense. I think of a hop as a node here, not a channel, tbh.

Reply:
Just saying that's my mental model, I think channel is valid as well. The docs seem accurate as-is to me since they speak in terms of `RouteHop`s, so they don't really make a call either way. We could clarify them further, but I think it would be clarifying pre-existing docs?

Reply:
> Just saying that's my mental model, I think channel is valid as well.

But the "channel" model is _not_ accurate in terms of how the docs are currently written. By saying "NOT INCLUDING our own", it would exclude the channel between ourselves and our peer, which is not the case.

So there's definitely ambiguity depending on how a hop is defined. We should at very least update the docs for `RotueHop` as "A hop in a route" doesn't provide much of a definition. In fact, I don't think we ever define what a hop _is_, either here or in `BlindedHop`, only what it is for. I think writing a definition in terms of HTLCs being sent would be clearer and remove ambiguity.

> The docs seem accurate as-is to me since they speak in terms of `RouteHop`s, so they don't really make a call either way. We could clarify them further, but I think it would be clarifying pre-existing docs?

Here, I was thinking more of implementation-level comments as this has been a source of confusion before. It wasn't clear to me the hop between the unblinded portion and the blinded portion is accounted for as an extra first "hop" in the blinded portion.